core/src/main/java/jenkins/security/ApiTokenProperty.java

```diff
@@ -356,6 +362,51 @@ public ApiTokenStats getTokenStats() {
         return tokenStats;
     }
 
+    /**
+     * It exposes /userToken/generate (requires POST).
+     * A token for the current user is returned as:
+     * <code>
+     *     {"status":"ok","data":{"tokenName":"test","tokenUuid":"a9edd227-3332-41a7-9baf-33091667b25b","tokenValue":"1142d457d1fe7628548f3215e680b31466"}}
+     * </code>
+     */
+    @Extension
+    public static class TokenGenerateAndGetREST extends CrumbExclusion implements RootAction {
+
+        @Override
+        public boolean process(HttpServletRequest request, HttpServletResponse response, FilterChain chain) throws IOException, ServletException {
+            String pathInfo = request.getPathInfo();
+            if (pathInfo != null && pathInfo.startsWith(getUrlName())) {
+                chain.doFilter(request, response);
+                return true;
+            }
+            return false;
+        }
+
+        @RequirePOST
+        public HttpResponse doGenerate(@QueryParameter String newTokenName) throws IOException {
+            // security checks are done by doGenerateNewToken
+            return new DescriptorImpl().doGenerateNewToken(User.current(), newTokenName);
```

Comment:
You need to validate `newTokenName`:
* Forbid them to create a token with an empty/blank name (return `400`) (of allow it but use a default token name)
* Does a token with that name already exist? If so, do you want to overwrite or return a `409 - conflict` status

Reply:
`DescriptorImpl().doGenerateNewToken` is already managing those cases.
* If the name is blank it generates something like `Token created on 2018-11-08T16:34:42.429+01:00`
* Tokens with the same name are allowed. I don't see why it should be restricted in this endpoint.